File: fn-tmdt/src/graphql/wallet.ts
import { gql } from '@apollo/client';

export interface WalletTransaction {
  id: string;
  walletId: string;
  amount: number;
  balanceBefore: number;
  balanceAfter: number;
  transactionType: string;
  status: string;
  referenceId: string | null;
  createdAt: string;
}

export interface Wallet {
  id: string;
  userId: string;
  balance: number;
  status: string;
  createdAt: string;
  updatedAt: string;
  transactions: WalletTransaction[];
}

export interface MyWalletData {
  myWallet: Wallet | null;
}

export const MY_WALLET_QUERY = gql`
  query MyWallet {
    myWallet {
      id
      userId
      balance
      status
      createdAt
      updatedAt
      transactions {
        id
        walletId
        amount
        balanceBefore
        balanceAfter
        transactionType
        status
        referenceId
        createdAt
      }
    }
  }
`;

export const REQUEST_WITHDRAWAL_MUTATION = gql`
  mutation RequestWithdrawal($amount: Float!, $bankDetails: String!) {
    requestWithdrawal(amount: $amount, bankDetails: $bankDetails) {
      id
      walletId
      amount
      balanceBefore
      balanceAfter
      transactionType
      status
      referenceId
      createdAt
    }
  }
`;

File: bk-tmdt/app/graphql/mutations/wallet.py
from decimal import Decimal
import json

import strawberry
from sqlalchemy.orm import Session
from sqlalchemy import select

from app.api.dependencies import get_db
from strawberry.types import Info
from app.graphql.types import WalletTransactionType, to_wallet_transaction_type
from app.models.entities import Wallet, WalletTransaction, WalletTransactionTypeEnum, WalletTransactionStatusEnum


@strawberry.type
class WalletMutation:
    @strawberry.mutation
    def request_withdrawal(
        self,
        info: Info,
        amount: float,
        bank_details: str,
    ) -> WalletTransactionType:
        user = info.context.get("current_user")
        if not user:
            raise Exception("Unauthorized")
        
        if amount < 50000:
            raise Exception("Minimum withdrawal amount is 50,000")

        db: Session = info.context.get("db")

        wallet = db.query(Wallet).filter(Wallet.user_id == user.id).with_for_update().first()
        if not wallet:
            raise Exception("Wallet not found")

        if wallet.balance < Decimal(str(amount)):
            raise Exception("Insufficient balance")

        # Deduct balance immediately
        balance_before = wallet.balance
        wallet.balance -= Decimal(str(amount))
        balance_after = wallet.balance

        # Create pending withdrawal transaction
        transaction = WalletTransaction(
            wallet_id=wallet.id,
            transaction_type=WalletTransactionTypeEnum.WITHDRAWAL,
            amount=Decimal(str(amount)),
            balance_before=balance_before,
            balance_after=balance_after,
            status=WalletTransactionStatusEnum.PENDING,
            reference_id=bank_details,
        )

        db.add(wallet)
        db.add(transaction)
        db.commit()
        db.refresh(transaction)

        return to_wallet_transaction_type(transaction)
